blog/content/post/2019-05-15-building-bareflank.md

```diff
@@ -0,0 +1,345 @@
+
+At the moment of writing this blog post, the newest tagged Bareflank version is
+`rc2.0.4`, but soon after it was released there was a big change in the
+structure of this project. Extended APIs (which are required for UEFI build and
+many other things) were merged into the main repository, earlier they were a
+project of its own. This made it difficult to maintain coherent interfaces
+between those two, usually one version of hypervisor was not compatible with
+some versions of extended APIs.
+
+As we really want to use extended APIs, in these series we will use commit
+`Merge pull request #733 from rianquinn/eapis_merge`. It was the latest commit
+that worked for UEFI platforms at the time of writing this article - code on
+`master` branch is not always working for all platforms; this is what tags are
+used for in this repository. There are also some examples merged into the
+hypervisor repository in later commits, earlier they were separate projects,
+just like the extended APIs. This ensures that they use the latest API, too.
+Perhaps we will look at them later in these series
+
+```
+git clone https://github.com/Bareflank/hypervisor.git
+cd hypervisor
+git co ba613e2c687f7042bac6886858cf6da3132a61d6
+cd ..
+mkdir build
+cp hypervisor/scripts/cmake/config/example_config.cmake config.cmake
+```
+
+This file needs to be changed in order to build Bareflank as a UEFI
+application. `config.cmake` is modified with almost every change, so it must be
+copied and set again if a different commit is used (also, clear `build`
+directory after such a change). Open this file in [your favourite editor](https://thomer.com/vi/vi.html)
+and change the following three lines:
+
+```
+set(ENABLE_BUILD_EFI OFF)
+(...)
+# set(OVERRIDE_VMM <name>)
+# set(OVERRIDE_VMM_TARGET <name>)
+```
+
+into this (note the lack of hashtags):
+
+```
+set(ENABLE_BUILD_EFI ON)
+(...)
+set(OVERRIDE_VMM integration_intel_x64_efi_test_efi)
+set(OVERRIDE_VMM_TARGET integration)
+```
+
+#### Fixing serial newline
+
+All Bareflank debug output is sent to the serial port (I/O port 0x3F8). Without
+heavy reverse engineering of OS it is impossible to print to the screen from
+VMM level. Only lines printed by UEFI entry point (either running at VM level
+or before entering VMX) to the hypervisor are visible also on the screen.
+
+Most clients for capturing data from UART assume that every line ends with
+`\r\n`, but Bareflank sends only '\n'. As a result, the cursor goes to the next
+line, but it does not return to the first column. This can be fixed with just
+one line added at the very beginning of the `write()` method in
+[serial_ns16550a.cpp](https://github.com/Bareflank/hypervisor/blob/ba613e2c687f7042bac6886858cf6da3132a61d6/bfvmm/src/debug/serial/serial_ns16550a.cpp#L234):
+
+```
+void
+serial_ns16550a::write(char c) const noexcept
+{
+    if (c == '\n') write('\r');
+
+    while (!is_transmit_empty())
+    { }
+
+    outb(0, static_cast<uint8_t>(c));
+}
+```
```

Comment:
@krystian-hebel typically in that situation I provide patch here - patch format is well understood for most technical people.

Reply:
Done.